File: mobile/mobile_central/test/services/modules/storefront/app/use_cases_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile_central/services/modules/storefront/app/use_cases.dart';
import 'package:mobile_central/services/modules/storefront/domain/entities.dart';
import 'package:mobile_central/services/modules/storefront/domain/ports.dart';
import 'package:mobile_central/shared/types/paginated_response.dart';

// --- Manual Mock ---

class MockStorefrontRepository implements IStorefrontRepository {
  final List<String> calls = [];

  PaginatedResponse<StorefrontProduct>? getCatalogResult;
  StorefrontProduct? getProductResult;
  Map<String, dynamic>? createOrderResult;
  PaginatedResponse<StorefrontOrder>? getOrdersResult;
  StorefrontOrder? getOrderResult;
  Map<String, dynamic>? registerResult;

  Exception? errorToThrow;

  GetCatalogParams? capturedCatalogParams;
  String? capturedProductId;
  int? capturedProductBusinessId;
  CreateStorefrontOrderDTO? capturedCreateOrderData;
  int? capturedCreateOrderBusinessId;
  GetOrdersParams? capturedOrdersParams;
  String? capturedOrderId;
  int? capturedOrderBusinessId;
  RegisterDTO? capturedRegisterData;

  @override
  Future<PaginatedResponse<StorefrontProduct>> getCatalog(GetCatalogParams? params) async {
    calls.add('getCatalog');
    capturedCatalogParams = params;
    if (errorToThrow != null) throw errorToThrow!;
    return getCatalogResult!;
  }

  @override
  Future<StorefrontProduct> getProduct(String id, {int? businessId}) async {
    calls.add('getProduct');
    capturedProductId = id;
    capturedProductBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return getProductResult!;
  }

  @override
  Future<Map<String, dynamic>> createOrder(CreateStorefrontOrderDTO data, {int? businessId}) async {
    calls.add('createOrder');
    capturedCreateOrderData = data;
    capturedCreateOrderBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return createOrderResult!;
  }

  @override
  Future<PaginatedResponse<StorefrontOrder>> getOrders(GetOrdersParams? params) async {
    calls.add('getOrders');
    capturedOrdersParams = params;
    if (errorToThrow != null) throw errorToThrow!;
    return getOrdersResult!;
  }

  @override
  Future<StorefrontOrder> getOrder(String id, {int? businessId}) async {
    calls.add('getOrder');
    capturedOrderId = id;
    capturedOrderBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return getOrderResult!;
  }

  @override
  Future<Map<String, dynamic>> register(RegisterDTO data) async {
    calls.add('register');
    capturedRegisterData = data;
    if (errorToThrow != null) throw errorToThrow!;
    return registerResult!;
  }
}

// --- Helpers ---

StorefrontProduct _makeProduct({String id = '1', String name = 'TestProduct'}) {
  return StorefrontProduct(
    id: id,
    name: name,
    description: 'desc',
    shortDescription: 'short',
    price: 10000.0,
    currency: 'COP',
    imageUrl: 'https://example.com/img.png',
    sku: 'SKU-1',
    stockQuantity: 10,
    category: 'Cat',
    brand: 'Brand',
    isFeatured: false,
    createdAt: '2026-01-01',
  );
}

StorefrontOrder _makeOrder({String id = '1'}) {
  return StorefrontOrder(
    id: id,
    orderNumber: 'ON-001',
    status: 'pending',
    totalAmount: 50000.0,
    currency: 'COP',
    createdAt: '2026-01-01',
    items: [],
  );
}

Pagination _makePagination() {
  return Pagination(
    currentPage: 1,
    perPage: 10,
    total: 1,
    lastPage: 1,
    hasNext: false,
    hasPrev: false,
  );
}

// --- Tests ---

void main() {
  late MockStorefrontRepository mockRepo;
  late StorefrontUseCases useCases;

  setUp(() {
    mockRepo = MockStorefrontRepository();
    useCases = StorefrontUseCases(mockRepo);
  });

  group('getCatalog', () {
    test('delegates to repository and returns result', () async {
      final expected = PaginatedResponse<StorefrontProduct>(
        data: [_makeProduct()],
        pagination: _makePagination(),
      );
      mockRepo.getCatalogResult = expected;
      final params = GetCatalogParams(page: 1, pageSize: 10);

      final result = await useCases.getCatalog(params);

      expect(result.data.length, 1);
      expect(result.data[0].name, 'TestProduct');
      expect(mockRepo.calls, ['getCatalog']);
      expect(mockRepo.capturedCatalogParams, params);
    });

    test('passes null params to repository', () async {
      mockRepo.getCatalogResult = PaginatedResponse<StorefrontProduct>(
        data: [],
        pagination: _makePagination(),
      );

      await useCases.getCatalog(null);

      expect(mockRepo.capturedCatalogParams, isNull);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Network error');

      expect(() => useCases.getCatalog(null), throwsException);
    });
  });

  group('getProduct', () {
    test('delegates to repository with correct id', () async {
      mockRepo.getProductResult = _makeProduct(id: 'p42', name: 'Found');

      final result = await useCases.getProduct('p42');

      expect(result.id, 'p42');
      expect(result.name, 'Found');
      expect(mockRepo.capturedProductId, 'p42');
      expect(mockRepo.calls, ['getProduct']);
    });

    test('passes businessId to repository', () async {
      mockRepo.getProductResult = _makeProduct();

      await useCases.getProduct('p1', businessId: 5);

      expect(mockRepo.capturedProductBusinessId, 5);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Not found');

      expect(() => useCases.getProduct('p1'), throwsException);
    });
  });

  group('createOrder', () {
    test('delegates to repository with correct data', () async {
      final dto = CreateStorefrontOrderDTO(
        items: [CreateStorefrontOrderItemDTO(productId: 'p1', quantity: 2)],
      );
      mockRepo.createOrderResult = {'id': 'ord-1', 'status': 'created'};

      final result = await useCases.createOrder(dto);

      expect(result['id'], 'ord-1');
      expect(mockRepo.capturedCreateOrderData, dto);
      expect(mockRepo.calls, ['createOrder']);
    });

    test('passes businessId to repository', () async {
      final dto = CreateStorefrontOrderDTO(items: []);
      mockRepo.createOrderResult = {};

      await useCases.createOrder(dto, businessId: 7);

      expect(mockRepo.capturedCreateOrderBusinessId, 7);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Creation failed');
      final dto = CreateStorefrontOrderDTO(items: []);

      expect(() => useCases.createOrder(dto), throwsException);
    });
  });

  group('getOrders', () {
    test('delegates to repository and returns result', () async {
      final expected = PaginatedResponse<StorefrontOrder>(
        data: [_makeOrder()],
        pagination: _makePagination(),
      );
      mockRepo.getOrdersResult = expected;
      final params = GetOrdersParams(page: 1);

      final result = await useCases.getOrders(params);

      expect(result.data.length, 1);
      expect(mockRepo.calls, ['getOrders']);
      expect(mockRepo.capturedOrdersParams, params);
    });

    test('passes null params to repository', () async {
      mockRepo.getOrdersResult = PaginatedResponse<StorefrontOrder>(
        data: [],
        pagination: _makePagination(),
      );

      await useCases.getOrders(null);

      expect(mockRepo.capturedOrdersParams, isNull);
    });
  });

  group('getOrder', () {
    test('delegates to repository with correct id', () async {
      mockRepo.getOrderResult = _makeOrder(id: 'ord-99');

      final result = await useCases.getOrder('ord-99');

      expect(result.id, 'ord-99');
      expect(mockRepo.capturedOrderId, 'ord-99');
      expect(mockRepo.calls, ['getOrder']);
    });

    test('passes businessId to repository', () async {
      mockRepo.getOrderResult = _makeOrder();

      await useCases.getOrder('ord-1', businessId: 3);

      expect(mockRepo.capturedOrderBusinessId, 3);
    });
  });

  group('register', () {
    test('delegates to repository with correct data', () async {
      final dto = RegisterDTO(
        name: 'Test',
        email: 'test@test.com',
        password: 'pass',
        businessCode: 'BC1',
      );
      mockRepo.registerResult = {'token': 'jwt-token'};

      final result = await useCases.register(dto);

      expect(result['token'], 'jwt-token');
      expect(mockRepo.capturedRegisterData, dto);
      expect(mockRepo.calls, ['register']);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Registration failed');
      final dto = RegisterDTO(
        name: 'T',
        email: 'e',
        password: 'p',
        businessCode: 'b',
      );

      expect(() => useCases.register(dto), throwsException);
    });
  });
}

File: front/central/src/services/modules/storefront/app/use-cases.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { StorefrontUseCases } from './use-cases';
import { IStorefrontRepository } from '../domain/ports';
import {
    StorefrontProduct,
    StorefrontOrder,
    StorefrontOrderItem,
    StorefrontClient,
    CreateStorefrontOrderDTO,
    CreateClientDTO,
    CreateClientResult,
    PaginatedResponse,
} from '../domain/types';

const makeProduct = (overrides: Partial<StorefrontProduct> = {}): StorefrontProduct => ({
    id: 'prod-001',
    name: 'Camiseta Negra',
    description: 'Camiseta de algodón 100%',
    short_description: 'Camiseta algodón',
    price: 45000,
    currency: 'COP',
    image_url: 'https://example.com/camiseta.png',
    sku: 'CAM-001',
    stock_quantity: 50,
    track_inventory: true,
    category: 'Ropa',
    brand: 'MarcaX',
    is_featured: true,
    created_at: '2026-01-01T00:00:00Z',
    ...overrides,
});

const makeOrderItem = (overrides: Partial<StorefrontOrderItem> = {}): StorefrontOrderItem => ({
    product_name: 'Camiseta Negra',
    quantity: 2,
    unit_price: 45000,
    total_price: 90000,
    ...overrides,
});

const makeOrder = (overrides: Partial<StorefrontOrder> = {}): StorefrontOrder => ({
    id: 'ord-001',
    order_number: '10001',
    status: 'pending',
    total_amount: 90000,
    currency: 'COP',
    created_at: '2026-03-24T00:00:00Z',
    items: [makeOrderItem()],
    ...overrides,
});

const makeClient = (overrides: Partial<StorefrontClient> = {}): StorefrontClient => ({
    id: 1,
    name: 'Juan Perez',
    email: 'juan@test.com',
    phone: '3001234567',
    dni: null,
    ...overrides,
});

const paginatedProducts: PaginatedResponse<StorefrontProduct> = {
    data: [makeProduct()],
    total: 1,
    page: 1,
    page_size: 10,
    total_pages: 1,
};

const paginatedOrders: PaginatedResponse<StorefrontOrder> = {
    data: [makeOrder()],
    total: 1,
    page: 1,
    page_size: 10,
    total_pages: 1,
};

const paginatedClients: PaginatedResponse<StorefrontClient> = {
    data: [makeClient()],
    total: 1,
    page: 1,
    page_size: 20,
    total_pages: 1,
};

function createMockRepository(): IStorefrontRepository {
    return {
        getCatalog: vi.fn(),
        getProduct: vi.fn(),
        createOrder: vi.fn(),
        getOrders: vi.fn(),
        getOrder: vi.fn(),
        createClient: vi.fn(),
        getClients: vi.fn(),
    } as unknown as IStorefrontRepository;
}

describe('StorefrontUseCases', () => {
    let repo: ReturnType<typeof createMockRepository>;
    let useCases: StorefrontUseCases;

    beforeEach(() => {
        repo = createMockRepository();
        useCases = new StorefrontUseCases(repo as unknown as IStorefrontRepository);
    });

    describe('getCatalog', () => {
        it('debería retornar el catálogo paginado de productos cuando el repositorio tiene éxito', async () => {
            vi.mocked(repo.getCatalog).mockResolvedValue(paginatedProducts);

            const result = await useCases.getCatalog({ page: 1, page_size: 10 });

            expect(result).toEqual(paginatedProducts);
            expect(repo.getCatalog).toHaveBeenCalledOnce();
            expect(repo.getCatalog).toHaveBeenCalledWith({ page: 1, page_size: 10 });
        });

        it('debería llamar al repositorio sin parámetros cuando no se pasan filtros', async () => {
            vi.mocked(repo.getCatalog).mockResolvedValue(paginatedProducts);

            await useCases.getCatalog();

            expect(repo.getCatalog).toHaveBeenCalledWith(undefined);
        });

        it('debería pasar filtros de búsqueda y categoría cuando se proporcionan', async () => {
            vi.mocked(repo.getCatalog).mockResolvedValue(paginatedProducts);

            await useCases.getCatalog({ search: 'camiseta', category: 'ropa', business_id: 1 });

            expect(repo.getCatalog).toHaveBeenCalledWith({ search: 'camiseta', category: 'ropa', business_id: 1 });
        });

        it('debería propagar el error cuando el repositorio falla', async () => {
            vi.mocked(repo.getCatalog).mockRejectedValue(new Error('Fallo de base de datos'));

            await expect(useCases.getCatalog()).rejects.toThrow('Fallo de base de datos');
        });
    });

    describe('getProduct', () => {
        it('debería retornar un producto por su ID', async () => {
            const product = makeProduct();
            vi.mocked(repo.getProduct).mockResolvedValue(product);

            const result = await useCases.getProduct('prod-001');

            expect(result).toEqual(product);
            expect(repo.getProduct).toHaveBeenCalledOnce();
            expect(repo.getProduct).toHaveBeenCalledWith('prod-001', undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const product = makeProduct();
            vi.mocked(repo.getProduct).mockResolvedValue(product);

            await useCases.getProduct('prod-001', 5);

            expect(repo.getProduct).toHaveBeenCalledWith('prod-001', 5);
        });

        it('debería propagar el error cuando el producto no existe', async () => {
            vi.mocked(repo.getProduct).mockRejectedValue(new Error('Producto no encontrado'));

            await expect(useCases.getProduct('no-existe')).rejects.toThrow('Producto no encontrado');
        });
    });

    describe('createOrder', () => {
        const dto: CreateStorefrontOrderDTO = {
            items: [{ product_id: 'prod-001', quantity: 2 }],
            notes: 'Entregar en la tarde',
        };

        it('debería crear una orden y retornar la respuesta del repositorio', async () => {
            const response = { message: 'Orden creada exitosamente' };
            vi.mocked(repo.createOrder).mockResolvedValue(response);

            const result = await useCases.createOrder(dto);

            expect(result).toEqual(response);
            expect(repo.createOrder).toHaveBeenCalledOnce();
            expect(repo.createOrder).toHaveBeenCalledWith(dto, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const response = { message: 'Orden creada exitosamente' };
            vi.mocked(repo.createOrder).mockResolvedValue(response);

            await useCases.createOrder(dto, 5);

            expect(repo.createOrder).toHaveBeenCalledWith(dto, 5);
        });

        it('debería propagar el error cuando la creación de la orden falla', async () => {
            vi.mocked(repo.createOrder).mockRejectedValue(new Error('Stock insuficiente'));

            await expect(useCases.createOrder(dto)).rejects.toThrow('Stock insuficiente');
        });
    });

    describe('getOrders', () => {
        it('debería retornar la lista paginada de órdenes cuando el repositorio tiene éxito', async () => {
            vi.mocked(repo.getOrders).mockResolvedValue(paginatedOrders);

            const result = await useCases.getOrders({ page: 1, page_size: 10 });

            expect(result).toEqual(paginatedOrders);
            expect(repo.getOrders).toHaveBeenCalledOnce();
            expect(repo.getOrders).toHaveBeenCalledWith({ page: 1, page_size: 10 });
        });

        it('debería llamar al repositorio sin parámetros cuando no se pasan filtros', async () => {
            vi.mocked(repo.getOrders).mockResolvedValue(paginatedOrders);

            await useCases.getOrders();

            expect(repo.getOrders).toHaveBeenCalledWith(undefined);
        });

        it('debería propagar el error cuando la consulta de órdenes falla', async () => {
            vi.mocked(repo.getOrders).mockRejectedValue(new Error('Servicio no disponible'));

            await expect(useCases.getOrders()).rejects.toThrow('Servicio no disponible');
        });
    });

    describe('getOrder', () => {
        it('debería retornar una orden por su ID', async () => {
            const order = makeOrder();
            vi.mocked(repo.getOrder).mockResolvedValue(order);

            const result = await useCases.getOrder('ord-001');

            expect(result).toEqual(order);
            expect(repo.getOrder).toHaveBeenCalledOnce();
            expect(repo.getOrder).toHaveBeenCalledWith('ord-001', undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const order = makeOrder();
            vi.mocked(repo.getOrder).mockResolvedValue(order);

            await useCases.getOrder('ord-001', 5);

            expect(repo.getOrder).toHaveBeenCalledWith('ord-001', 5);
        });

        it('debería propagar el error cuando la orden no existe', async () => {
            vi.mocked(repo.getOrder).mockRejectedValue(new Error('Orden no encontrada'));

            await expect(useCases.getOrder('no-existe')).rejects.toThrow('Orden no encontrada');
        });
    });

    describe('createClient', () => {
        const dto: CreateClientDTO = {
            name: 'Juan Perez',
            email: 'juan@test.com',
            phone: '3001234567',
        };

        it('debería crear un cliente y devolver la contrasena temporal generada', async () => {
            const response: CreateClientResult = { client: makeClient(), temp_password: 'aB3xY9zK1' };
            vi.mocked(repo.createClient).mockResolvedValue(response);

            const result = await useCases.createClient(dto);

            expect(result).toEqual(response);
            expect(repo.createClient).toHaveBeenCalledOnce();
            expect(repo.createClient).toHaveBeenCalledWith(dto, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const response: CreateClientResult = { client: makeClient() };
            vi.mocked(repo.createClient).mockResolvedValue(response);

            await useCases.createClient(dto, 5);

            expect(repo.createClient).toHaveBeenCalledWith(dto, 5);
        });

        it('debería propagar el error cuando el email ya existe', async () => {
            vi.mocked(repo.createClient).mockRejectedValue(new Error('Email ya registrado'));

            await expect(useCases.createClient(dto)).rejects.toThrow('Email ya registrado');
        });
    });

    describe('getClients', () => {
        it('debería retornar la lista paginada de clientes', async () => {
            vi.mocked(repo.getClients).mockResolvedValue(paginatedClients);

            const result = await useCases.getClients({ page: 1, page_size: 20 });

            expect(result).toEqual(paginatedClients);
            expect(repo.getClients).toHaveBeenCalledWith({ page: 1, page_size: 20 });
        });

        it('debería propagar el error cuando la consulta falla', async () => {
            vi.mocked(repo.getClients).mockRejectedValue(new Error('Servicio no disponible'));

            await expect(useCases.getClients()).rejects.toThrow('Servicio no disponible');
        });
    });
});
